framework/src/meshgenerators/Poly2TriMeshGenerator.C

```diff
@@ -0,0 +1,242 @@
+
+      if (cross_prod(2) <= 0)
+        mooseError("Inverted element found in triangulation.\n"
+                   "Laplacian smoothing can create these at reentrant corners; disable it?");
+    }
+
+  const bool use_binary_search = (_algorithm == "BINARY");
+
+  for (auto hole_i : index_range(_hole_ptrs))
+  {
+    if (hole_i < _stitch_holes.size() && _stitch_holes[hole_i])
+    {
+      // The hole mesh was specified by the user, so it could have
+      // any BCID or no BCID or any combination of BCIDs on its
+      // outer boundary.  Let's just pick out our own temporary
+      // BCIDs to use for stitching.
+
+      // The new triangulation assigns BCID i+1 to hole i ... but
+      // we can't even use this for mesh stitching, because we
+      // can't be sure it isn't also already in use on the hole's
+      // mesh and so we won't be able to safely clear it
+      // afterwards.
+
+      UnstructuredMesh & hole_mesh = dynamic_cast<UnstructuredMesh &>(*_hole_ptrs[hole_i]->get());
+      auto & hole_boundary_info = hole_mesh.get_boundary_info();
+      const std::set<boundary_id_type> & local_hole_bcids = hole_boundary_info.get_boundary_ids();
+
+      boundary_id_type new_hole_bcid = 0;
+      if (!local_hole_bcids.empty())
+        new_hole_bcid = *local_hole_bcids.rbegin();
+      hole_mesh.comm().max(new_hole_bcid);
+
+      ++new_hole_bcid;
+      const boundary_id_type inner_bcid = new_hole_bcid + 1;
+
+      // It would have been nicer for MeshedHole to add the BCID
+      // itself, but we want MeshedHole to work with a const mesh.
+      // We'll still use MeshedHole, for its code distinguishing
+      // outer boundaries from inner boundaries on a
+      // hole-with-holes.
+
+      TriangulatorInterface::MeshedHole mh{hole_mesh};
+
+      // We have to translate from MeshedHole points to mesh
+      // sides.
+      std::unordered_map<Point, Point> next_hole_boundary_point;
+      const int np = mh.n_points();
+      for (auto pi : make_range(1, np))
+        next_hole_boundary_point[mh.point(pi - 1)] = mh.point(pi);
+      next_hole_boundary_point[mh.point(np - 1)] = mh.point(0);
+
+      int found_hole_sides = 0;
+      for (auto elem : hole_mesh.element_ptr_range())
+      {
+        auto ns = elem->n_sides();
+        for (auto s : make_range(ns))
+        {
+          // Remember that the point ordering should be
+          // reversed since we're thinking of it as an outer
+          // boundary rather than an inner here.
+          auto it_s = next_hole_boundary_point.find(elem->point((s + 1) % ns));
+          if (it_s != next_hole_boundary_point.end())
+            if (it_s->second == elem->point(s))
+            {
+              hole_boundary_info.add_side(elem, s, new_hole_bcid);
+              ++found_hole_sides;
+            }
+        }
+      }
+      mooseAssert(found_hole_sides == np, "Failed to find full outer boundary of meshed hole");
```

Comment:
how fool proof is this? Do we expect users to hit this in difficult cases?

Reply:
in fact what if the user came in with a mesh that had a different number of outer points than what we used to mesh the hole outside?

Reply:
If I expected difficult cases to trigger it I'd have made it a mooseError; this should only trigger if we have an internal bug ... which is thinkable enough that I didn't want want to omit the assert entirely.

The hole is purely defined by the hole mesh ... but I suppose we could end up with an inconsistency before stitching if a user requested a hole be stitched but didn't disallow the hole from being automatically refined.  I'll add an error if that's ever inconsistent.

Reply:
ok and what about if the hole has a hole?
or what if it s a 2D mesh?

Reply:
Holes with holes are supported (MeshedHole only finds an outer boundary, so then the inner boundary doesn't get the temporary stitching BCID assigned).  Holes defined by 2D meshes are supported - that's where the stitching option becomes useful.